### PR TITLE
From cmssw 8 0 8

### PR DESCRIPTION
**Pull request by akshanshs**
BadPFmuon and BadChHadron filter update

### CODE REVIEW

**Comment by slava77** [comment]
@akshanshs 
Please change the title of the PR
I can suggest "BadPFmuon and BadChHadron filter update" (there is an "Edit" button on the upper right on the web interface).
Also, please edit the PR description, the first text entry after the PR title  (click pencil-like icon to edit) and provide more details of what is in this PR: a presentation in a POG meeting and some more details about the PR.

File: RecoMET/METFilters/python/BadChargedCandidateFilter_cfi.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import FWCore.ParameterSet.Config as cms
+
+BadChargedCandidateFilter = cms.EDFilter(
+    "BadChargedCandidateFilter",
+    PFCandidates  = cms.InputTag("particleFlow"),   # Collection to test
+    muons  = cms.InputTag("muons"),   # Collection to test
+    taggingMode   = cms.bool(False),
+    debug         = cms.bool(True),
```
Comment (slava77) [inline RecoMET/METFilters/python/BadChargedCandidateFilter_cfi.py]: this has to be "False" in the release version.

File: RecoMET/METFilters/plugins/BadChargedCandidateFilter.cc [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            if (not ( abs(pfCandidate.pdgId()) == 211) ) continue;
+            float dr = deltaR( innerMuonTrack->eta(), innerMuonTrack->phi(), pfCandidate.eta(), pfCandidate.phi() );
+            float dpt = ( pfCandidate.pt() - innerMuonTrack->pt())/(0.5*(innerMuonTrack->pt() + pfCandidate.pt()));
+            if ( (debug_)  and (dr<0.5) ) cout<<" pt(it) "<<innerMuonTrack->pt()<<" candidate "<<pfCandidate.pt()<<" dr "<< dr
+                <<" dpt "<<dpt<<endl;
+            // require similar pt ( one sided ) and small dR
+            if ( ( deltaR( innerMuonTrack->eta(), innerMuonTrack->phi(), pfCandidate.eta(), pfCandidate.phi() ) < maxDR_ ) 
+                and ( ( pfCandidate.pt() - innerMuonTrack->pt())/(0.5*(innerMuonTrack->pt() + pfCandidate.pt())) > minPtDiffRel_ ) ) {
+                    foundBadChargedCandidate = true;
+                    cout <<"found bad track!"<<endl;
```
Comment (slava77) [inline RecoMET/METFilters/plugins/BadChargedCandidateFilter.cc]: this should be at least protected by "if debug_"

File: RecoMET/METFilters/plugins/BadChargedCandidateFilter.cc [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  typedef View<reco::Muon> MuonView;
+  Handle<MuonView> muons;
+  iEvent.getByToken(tokenMuons_,muons);
+
+  bool foundBadChargedCandidate = false;
+
+  for ( unsigned i=0; i < muons->size(); ++i ) { // loop over all muons
+
+    const reco::Muon & muon = (*muons)[i];
+    if (debug_) cout<<"BadChargedCandidate test"<<endl;
```
Comment (slava77) [inline RecoMET/METFilters/plugins/BadChargedCandidateFilter.cc]: Note that "cout" is not thread safe and will produce garbled output if the code is running multithreaded.
Better change to "LogInfo" (not printed by MessageLogger by default) or "LogPrint" (printed by MessageLogger by default) to use thread safe version.
We typically allow the pattern ``` if (debug) cout << ... ```, but it should be understood that it has issues and will not allow debugging of a job running with multiple threads.

File: RecoMET/METFilters/plugins/BadPFMuonFilter.cc [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+  bool foundBadPFMuon = false;
+
+  for ( unsigned i=0; i < muons->size(); ++i ) { // loop over all muons
+    
+    const reco::Muon & muon = (*muons)[i];
+
+    reco::TrackRef innerMuonTrack = muon.innerTrack();
+    
+    cout << "PF filter muon:" << i << endl ;
```
Comment (slava77) [inline RecoMET/METFilters/plugins/BadPFMuonFilter.cc]: exposed cout is not allowed.
Use "if (debug" or change to LogTrace

File: RecoMET/METFilters/plugins/BadPFMuonFilter.cc [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            if (debug_) cout<<"Skipping this muon because inner track pt."<<endl; 
+       continue;
+    }
+
+    if ( innerMuonTrack->quality(reco::TrackBase::highPurity) ) { 
+      //      if (debug_) cout<<"Skipping this muon because inner track is high purity."<<endl; 
+      // continue;
+    }
+    
+       cout << "SegmentCompatibility :"<< muon::segmentCompatibility(muon) << "pt:" << innerMuonTrack->pt() << endl;
```
Comment (slava77) [inline RecoMET/METFilters/plugins/BadPFMuonFilter.cc]: if debug

File: RecoMET/METFilters/test/test_badChargedCandidateFilter.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    'PoolSource',
+    #RECO
+    #fileNames = cms.untracked.vstring('root://eoscms.cern.ch//store/group/phys_jetmet/MetScanners/bobak_pickevents.root')
+    #fileNames = cms.untracked.vstring('file:/afs/cern.ch/user/m/mzeinali/public/skim_9.root')
+    #fileNames = cms.untracked.vstring( 'file:lowMet_failBadChargedHadronFilter.root')
+    #miniAOD
+    #fileNames = cms.untracked.vstring('root://eoscms.cern.ch//store/group/phys_jetmet/MetScanners/bobak_pickevents_miniAOD.root ')
+    #fileNames = cms.untracked.vstring('root://eoscms.cern.ch//store/mc/RunIISpring16MiniAODv1/SMS-T1tttt_mGluino-1500_mLSP-100_TuneCUETP8M1_13TeV-madgraphMLM-pythia8/MINIAODSIM/PUSpring16_80X_mcRun2_asymptotic_2016_v3-v1/60000/26C76166-0FFE-E511-BA96-0025905D1D60.root')
+    #fileNames = cms.untracked.vstring( "file:/afs/cern.ch/user/s/schoef/eos/cms/store/mc/RunIISpring16MiniAODv1/SMS-T1tttt_mGluino-1500_mLSP-100_TuneCUETP8M1_13TeV-madgraphMLM-pythia8/MINIAODSIM/PUSpring16_80X_mcRun2_asymptotic_2016_v3-v1/60000/26C76166-0FFE-E511-BA96-0025905D1D60.root")
+    fileNames = cms.untracked.vstring( 'file:/data/rschoefbeck/local/SMS-T1tttt_mGluino-1500_mLSP-100_TuneCUETP8M1_13_MINIAODSIM.root')
```
Comment (slava77) [inline RecoMET/METFilters/test/test_badChargedCandidateFilter.py]: please cleanup files with afs paths and leave something from eos or in global space "/store/... "

**Comment by mariadalfonso** [comment]
@slava77 
These PR are still under works w/ MET group, so we should close the PR

**Pull request closed without merging**